tests/test_trusted_metadata_set.py

```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_snapshot(modified_snapshot_data)
+
+    def test_update_snapshot_after_succesfull_update_new_snapshot_meta_version_different(self):
+        self._setup_update_snapshot_after_successful_update_test()
+        # snapshot.meta["project1"].version != new_snapshot.meta["project1"].version
+        for meta_file_path in self.trusted_set.snapshot.signed.meta.keys():
+            self.trusted_set.snapshot.signed.meta[meta_file_path].version = 2
+        with self.assertRaises(exceptions.BadVersionNumberError):
+            self.trusted_set.update_snapshot(self.metadata["snapshot"])
+
+    def test_update_snapshot_after_succesfull_expired_new_snapshot(self):
+        self._setup_update_snapshot_after_successful_update_test()
+        # new_snapshot has expired
+        snapshot_obj = Metadata.from_bytes(self.metadata["snapshot"])
+        snapshot_obj.signed.expires = datetime.strptime(
+            "1970-01-01T00:00:00Z", "%Y-%m-%dT%H:%M:%SZ"
+        )
+        self._sign_modified_obj("snapshot", snapshot_obj)
+        modified_snapshot_data = json.dumps(snapshot_obj.to_dict()).encode()
+        timestamp_meta = self.trusted_set.timestamp.signed.meta["snapshot.json"]
+        true_hashes = timestamp_meta.hashes or {}
+        modified_hashes = self._calculate_modified_hashes(
+            true_hashes, modified_snapshot_data
+        )
+        self.trusted_set.timestamp.signed.meta["snapshot.json"].hashes = modified_hashes
+        # remove length so it doesn't intervene the validation.
+        self.trusted_set.timestamp.signed.meta["snapshot.json"].length = None
+        with self.assertRaises(exceptions.ExpiredMetadataError):
+            self.trusted_set.update_snapshot(modified_snapshot_data)
+
+
+    def setup_update_targets_test(self):
+        self.trusted_set.root_update_finished()
+        self.trusted_set.update_timestamp(self.metadata["timestamp"])
+        self.trusted_set.update_snapshot(self.metadata["snapshot"])
+
+    def test_update_targets_no_meta_in_snapshot(self):
+        self.setup_update_targets_test()
+        # remove meta information with information about targets from snapshot
+        self.trusted_set.snapshot.signed.meta = {}
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_targets(self.metadata["targets"])
+
+    def test_update_targets_hash_different_than_snapshot_meta_hash(self):
+        self.setup_update_targets_test()
+        # observed_hash != stored hash in snapshot meta for targets
+        true_hashes = {}
+        for target_path, meta_file in self.trusted_set.snapshot.signed.meta.items():
+            true_hashes[target_path] = meta_file.hashes
+            self.trusted_set.snapshot.signed.meta[target_path].hashes = {"sha256": "b"}
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_targets(self.metadata["targets"])
```

Comment:
true_hashes is unused

Also I think mentioning hashes in the test name is maybe not what we want: TrustedMetadataSet is not interested in whether the issue is hashes, version or length: we're testing that TrustedMetadataSet functions correctly when MetaFile verification fails (for any reason)

Reply:
Yes, but I want to be specific about what's wrong with this particular test because we have around 15 of them.

Reply:
> true_hashes is unused

Reply:
Thanks, I missed that.